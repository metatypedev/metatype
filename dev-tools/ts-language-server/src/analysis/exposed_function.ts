import { Parser, TypeScript } from "../parser.ts";

type ExposedFunction = {
  name: string;
  node: Parser.SyntaxNode;
  runtime: Parser.SyntaxNode;
  generator: string;
  generatorArgs: Parser.SyntaxNode;
  apply?: Parser.SyntaxNode | null;
  policy?: Parser.SyntaxNode | null;
};

type MethodCall = {
  object: Parser.SyntaxNode;
  method: string;
  arguments: Parser.SyntaxNode;
};

export function asMethodCall(node: Parser.SyntaxNode): MethodCall | null {
  if (node.type !== "call_expression") {
    return null;
  }
  const fn = node.childForFieldName("function")!;
  if (fn.type !== "member_expression") {
    return null;
  }
  const object = fn.childForFieldName("object")!;
  const property = fn.childForFieldName("property")!;
  if (property.type !== "property_identifier") {
    return null;
  }
  const argumentsNode = node.childForFieldName("arguments")!;
  if (argumentsNode.type !== "arguments") {
    return null;
  }
  return {
    object: object,
    method: property.text,
    arguments: argumentsNode,
  };
}

/**
 * Analyze an expose expression
 *
 * TODO: variable resolution (use the definition subtree)
 *   to be implemented in asMethodCall
 *   use default LSP to find definition
 *
 * Example:
 * database.findMany(users).apply({ where: { id: 1 } }).withPolicy(policy)
 * - runtime: database
 * - generator: findMany
 * - generatorArgs: users
 * - apply: { where: { id: 1 } }
 * - policy: policy
 */
export function analyzeExposeExpression(
  node: Parser.SyntaxNode,
): Omit<ExposedFunction, "name"> {
  let methodCall = asMethodCall(node);
  if (methodCall === null) {
    throw new Error("expected method call");
  }
  let policy: Parser.SyntaxNode | null = null;
  let apply: Parser.SyntaxNode | null = null;
  if (methodCall.method === "withPolicy") {
    policy = methodCall.arguments;
    methodCall = asMethodCall(methodCall.object);
    if (methodCall === null) {
      throw new Error("expected method call");
    }
  }

  // TODO what if apply is a generator name??
  if (methodCall.method === "apply") {
    apply = methodCall.arguments;
    methodCall = asMethodCall(methodCall.arguments);
    if (methodCall === null) {
      throw new Error("expected method call");
    }
  }

  const runtime = methodCall.object;
  const generator = methodCall.method;
  const generatorArgs = methodCall.arguments;

  return {
    node,
    runtime,
    generator,
    generatorArgs,
    apply,
    policy,
  };
}
