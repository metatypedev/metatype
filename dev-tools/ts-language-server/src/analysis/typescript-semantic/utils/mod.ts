import { Parser } from "../../../parser.ts";

export type MethodCall = {
  object: Parser.SyntaxNode;
  method: Parser.SyntaxNode;
  arguments: Parser.SyntaxNode;
};

export function asMethodCall(node: Parser.SyntaxNode): MethodCall | null {
  if (node.type !== "call_expression") {
    return null;
  }
  const fn = node.childForFieldName("function")!;
  if (fn.type !== "member_expression") {
    return null;
  }
  const object = fn.childForFieldName("object")!;
  const property = fn.childForFieldName("property")!;
  if (property.type !== "property_identifier") {
    return null;
  }
  const argumentsNode = node.childForFieldName("arguments")!;
  if (argumentsNode.type !== "arguments") {
    return null;
  }
  return {
    object: object,
    method: property,
    arguments: argumentsNode,
  };
}
