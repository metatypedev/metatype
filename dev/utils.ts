// Copyright Metatype OÃœ, licensed under the Elastic License 2.0.
// SPDX-License-Identifier: Elastic-2.0

import { dirname, fromFileUrl, resolve, yaml } from "./deps.ts";

export const projectDir = resolve(
  dirname(fromFileUrl(import.meta.url)),
  "..",
);

export async function runOrExit(
  cmd: string[],
  cwd?: string,
  env: Record<string, string> = {},
) {
  // console.error("shell: ", cmd.join(" "), { env, cwd });
  const p = new Deno.Command(cmd[0], {
    args: cmd.slice(1),
    cwd,
    // stdout: "piped",
    // stderr: "piped",
    env: { ...Deno.env.toObject(), ...env },
  }).spawn();

  // await p.stdout.pipeTo(Deno.stdout.writable, { preventClose: true });
  // await p.stderr.pipeTo(Deno.stderr.writable, { preventClose: true });

  const { code, success } = await p.status;
  if (!success) {
    Deno.exit(code);
  }
}

export function relPath(path: string) {
  let clean = path.replace(projectDir, "");
  if (clean.startsWith("/")) {
    clean = clean.slice(1);
  }
  return clean;
}

interface Lockfile {
  [channel: string]: {
    files: Record<string, string[]>;
    lines: Record<string, Record<string, string>>;
    lock: Record<string, string>;
  };
}

export const lockfileUrl = resolve(projectDir, "dev/lock.yml");

export async function getLockfile() {
  const file = await Deno.readTextFile(lockfileUrl);
  return yaml.parse(
    file,
  ) as Lockfile;
}

export type Cursor = {
  start: number;
  end: number;
  length: number;
  match: string;
};

export function upperFirst(str: string) {
  return str.charAt(0).toUpperCase() + str.substring(1);
}

export function camelCase(str: string) {
  return str
    .split(/_+/g)
    .map((chunk, idx) => idx > 0 ? upperFirst(chunk) : chunk)
    .join("");
}

/**
 * Enhanced `indexOf` with regex support and position information
 */
export function nextMatch(
  text: string,
  word: string | RegExp,
  pos = 0,
): Cursor | null {
  if (word instanceof RegExp) {
    const searchPos = Math.min(text.length, pos);
    const nextText = text.substring(searchPos);
    const res = word.exec(nextText);
    word.lastIndex = 0; // always reset (js!)
    return res
      ? {
        match: res[0],
        start: searchPos + res.index,
        end: searchPos + res.index + res[0].length - 1,
        length: res[0].length,
      }
      : null;
  }
  const start = text.indexOf(word, pos);
  return start >= 0
    ? {
      start,
      end: start + word.length - 1,
      match: word,
      length: word.length,
    }
    : null;
}

/**
 * Determine all indexOf with position information
 */
export function findCursors(
  text: string,
  word: string | RegExp,
): Array<Cursor> {
  const matches = [] as Array<Cursor>;

  let cursor = 0;
  while (true) {
    const res = nextMatch(text, word, cursor);
    if (res != null) {
      cursor = res.end;
      matches.push(res);
    } else {
      break;
    }
  }

  return matches;
}
