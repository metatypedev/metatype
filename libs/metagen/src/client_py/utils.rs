// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use heck::*;
use once_cell::sync::Lazy;

// sourced from
// keyword library https://docs.python.org/3/library/keyword.html
static _KEYWORDS: Lazy<std::collections::HashSet<&'static str>> = Lazy::new(|| {
    [
        "False", "None", "True", "and", "as", "assert", "async", "await", "break", "class",
        "continue", "def", "del", "elif", "else", "except", "finally", "for", "from", "global",
        "if", "import", "in", "is", "lambda", "nonlocal", "not", "or", "pass", "raise", "return",
        "try", "while", "with", "yield",
    ]
    .into_iter()
    .collect()
});

pub fn normalize_type_title(title: &str) -> String {
    static RE: Lazy<regex::Regex> =
        Lazy::new(|| regex::Regex::new(r"^(?<startd>\d+)(?<rest>.*)").unwrap());

    // TODO: clean up non valid chars

    // clean out underscores at start/end
    let title = title.trim_matches('_');
    // move any numbers at start to end
    let title = title.to_pascal_case();
    let title = RE.replace(&title, "$rest$startd");
    title.to_string()
}

/// This assumes title was already through [normalize_type_title]
pub fn normalize_struct_prop_name(title: &str) -> String {
    title.to_snek_case()
    // if KEYWORDS.contains(&title[..]) {
    //     format!("{title}_")
    // } else {
    //     title
    // }
}
