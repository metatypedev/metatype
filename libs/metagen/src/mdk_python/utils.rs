// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use indexmap::IndexMap;
use serde::Serialize;

#[derive(Debug, Eq, Hash, PartialEq, Clone, Serialize)]
pub struct TypeGenerated {
    /// Type representation: int, str, class name, ..
    pub hint: String,
    /// Source-code (for structs only)
    pub def: Option<String>,
}

impl From<String> for TypeGenerated {
    fn from(value: String) -> Self {
        Self {
            hint: value.to_string(),
            def: None,
        }
    }
}

#[derive(Debug, Clone)]
struct Class {
    priority: u32,
    type_generated: Option<TypeGenerated>,
}

#[derive(Clone)]
pub struct Memo {
    map: IndexMap<String, Class>,
    priority_weight: u32,
}

impl Memo {
    pub fn new() -> Self {
        Self {
            map: IndexMap::new(),
            priority_weight: 1,
        }
    }

    /// Insert `v` at `k`, if already present, this will also increase the priority by the current weight
    pub fn insert(&mut self, k: String, v: TypeGenerated) {
        if !self.is_allocated(&k) {
            self.allocate(k.clone());
        }
        let old = self.map.get(&k).unwrap();
        self.map.insert(
            k,
            Class {
                priority: old.priority + self.priority_weight,
                type_generated: Some(v),
            },
        );
    }

    /// Allocate for `k` and set priority to 0, if a place is already allocated this will do nothing
    pub fn allocate(&mut self, k: String) {
        if !self.is_allocated(&k) {
            self.map.insert(
                k,
                Class {
                    priority: 0,
                    type_generated: None,
                },
            );
        }
    }

    pub fn is_allocated(&self, k: &str) -> bool {
        self.map.contains_key(k)
    }

    pub fn types_in_order(&self) -> Vec<TypeGenerated> {
        let mut values = self
            .map
            .values()
            .filter(|c| c.type_generated.is_some())
            .collect::<Vec<_>>();

        // if typing dataclass A depends on dataclass B, then B must be generated first
        values.sort_by(|a, b| b.priority.cmp(&a.priority));

        values
            .iter()
            .filter_map(|c| c.type_generated.as_ref().cloned())
            .collect()
    }

    pub fn incr_weight(&mut self) {
        self.priority_weight += 1;
    }

    pub fn decr_weight(&mut self) {
        if let Some(value) = self.priority_weight.checked_sub(1) {
            self.priority_weight = value;
        } else {
            panic!("invalid state: priority weight overflowed")
        }
    }

    pub fn merge_with(&mut self, other: Memo) {
        for (k, v) in other.map.into_iter() {
            if let Some(gen) = v.type_generated {
                self.insert(k, gen);
            }
        }
    }
}
