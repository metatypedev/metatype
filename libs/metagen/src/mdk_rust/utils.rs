// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use heck::*;
use once_cell::sync::Lazy;

static KEYWORDS: Lazy<std::collections::HashSet<&'static str>> = Lazy::new(|| {
    [
        // strict keywords
        "as", "break", "const", "continue", "crate", "else", "enum", "extern", "false", "fn", "for",
        "if", "impl", "in", "let", "loop", "match", "mod", "move", "mut", "pub", "ref", "return",
        "self", "Self", "static", "struct", "super", "trait", "true", "type", "unsafe", "use",
        "where", "while", //
        // 2018+ keywords
        "async", "await", "dyn", //
        // reserved keywords
        "abstract", "become", "box", "do", "final", "macro", "override", "priv", "typeof",
        "unsized", "virtual", "yield", //
        // 2018+ reserved keywords
        "try",
    ]
    .into_iter()
    .collect()
});

pub fn normalize_type_title(title: &str) -> String {
    static RE: Lazy<regex::Regex> =
        Lazy::new(|| regex::Regex::new(r"^(?<startd>\d+)(?<rest>.*)").unwrap());

    // TODO: clean out rust keywords
    // TODO: clean up non valid chars

    // clean out underscores at start/end
    let title = title.trim_matches('_');
    // move any numbers at start to end
    let title = title.to_pascal_case();
    let title = RE.replace(&title, "$rest$startd");
    if KEYWORDS.contains(&title[..]) {
        format!("r#{title}")
    } else {
        title.to_string()
    }
}

pub fn normalize_struct_prop_name(title: &str) -> String {
    let title = title.to_snek_case();
    if KEYWORDS.contains(&title[..]) {
        format!("r#{title}")
    } else {
        title
    }
}
