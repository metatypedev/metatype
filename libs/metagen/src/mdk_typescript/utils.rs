// Copyright Metatype OÃœ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

use heck::*;
use once_cell::sync::Lazy;

// sourced from
// https://github.com/microsoft/TypeScript/blob/ef514af2675389d38c793d6cc1945486c367e6fa/src/compiler/types.ts#L140
static KEYWORDS: Lazy<std::collections::HashSet<&'static str>> = Lazy::new(|| {
    [
        "break",
        "case",
        "catch",
        "class",
        "const",
        "continue",
        "debugger",
        "default",
        "delete",
        "do",
        "else",
        "enum",
        "export",
        "extends",
        "false",
        "finally",
        "for",
        "function",
        "if",
        "import",
        "in",
        "instanceof",
        "new",
        "null",
        "return",
        "super",
        "switch",
        "this",
        "throw",
        "true",
        "try",
        "typeof",
        "var",
        "void",
        "while",
        "with",
        // strict mode reserved words
        "implements",
        "interface",
        "let",
        "package",
        "private",
        "protected",
        "public",
        "static",
        "yield",
        // contextual keywords
        "abstract",
        "accessor",
        "as",
        "asserts",
        "assert",
        "any",
        "async",
        "await",
        "boolean",
        "constructor",
        "declare",
        "get",
        "infer",
        "intrinsic",
        "is",
        "keyof",
        "module",
        "namespace",
        "never",
        "out",
        "readonly",
        "require",
        "number",
        "object",
        "satisfies",
        "set",
        "string",
        "symbol",
        "type",
        "undefined",
        "unique",
        "unknown",
        "using",
        "from",
        "global",
        "bigint",
        "override",
        "of",
    ]
    .into_iter()
    .collect()
});

pub fn normalize_type_title(title: &str) -> String {
    static RE: Lazy<regex::Regex> =
        Lazy::new(|| regex::Regex::new(r"^(?<startd>\d+)(?<rest>.*)").unwrap());

    // TODO: clean up non valid chars

    // clean out underscores at start/end
    let title = title.trim_matches('_');
    // move any numbers at start to end
    let title = title.to_pascal_case();
    let title = RE.replace(&title, "$rest$startd");
    title.to_string()
}

/// This assumes title was already through [normalize_type_title]
pub fn normalize_struct_prop_name(title: &str) -> String {
    if KEYWORDS.contains(title) {
        format!("\"{title}\"")
    } else {
        title.to_string()
    }
}
