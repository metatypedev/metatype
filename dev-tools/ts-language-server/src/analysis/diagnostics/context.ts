// import { Diagnostic } from "vscode-languageserver";
import { Diagnostic, DiagnosticSeverity } from "vscode-languageserver/types";
import { ScopeManager } from "../typescript-semantic/scope.ts";
import { Parser } from "../../parser.ts";
import { ExposedFunction } from "../exposed_function.ts";
import { TgTypeStruct } from "../typescript-semantic/semantic-node.ts";

export class ModuleDiagnosticsContext {
  symbolRegistry: ScopeManager;
  diagnostics: Diagnostic[] = [];

  constructor(rootNode: Parser.SyntaxNode, private uri: string) {
    this.symbolRegistry = new ScopeManager(rootNode);
  }

  #pushDiagnostic(
    severity: DiagnosticSeverity,
    node: Parser.SyntaxNode,
    message: string,
  ) {
    this.diagnostics.push({
      severity,
      message,
      range: {
        start: {
          line: node.startPosition.row,
          character: node.startPosition.column,
        },
        end: {
          line: node.endPosition.row,
          character: node.endPosition.column,
        },
      },
      source: "typegraph",
    });
  }

  checkExposedFunction(exposedFunction: ExposedFunction) {
    const inputType = exposedFunction.inputType;
    if (!(inputType instanceof TgTypeStruct)) {
      this.error(
        exposedFunction.node,
        `Input type of exposed function must be a struct, but got ${inputType}`,
      );
    }
  }

  error(node: Parser.SyntaxNode, message: string) {
    this.#pushDiagnostic(DiagnosticSeverity.Error, node, message);
  }

  warn(node: Parser.SyntaxNode, message: string) {
    this.#pushDiagnostic(DiagnosticSeverity.Warning, node, message);
  }

  info(node: Parser.SyntaxNode, message: string) {
    this.#pushDiagnostic(DiagnosticSeverity.Information, node, message);
  }

  hint(node: Parser.SyntaxNode, message: string) {
    this.#pushDiagnostic(DiagnosticSeverity.Hint, node, message);
  }
}
