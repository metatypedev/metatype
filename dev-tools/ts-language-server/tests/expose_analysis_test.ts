import { assertEquals, assertObjectMatch } from "std/assert/mod.ts";
import { join } from "std/path/mod.ts";
import { testDir } from "./utils.ts";
import Parser from "npm:web-tree-sitter";
import {
  findTypegraphDefinitions,
  TypegraphDefinition,
} from "../src/parser.ts";
import { analyzeExposeExpression } from "../src/analysis/exposed_function.ts";
import { ScopeManager } from "../src/analysis/typescript-semantic/scope.ts";
import { ModuleDiagnosticsContext } from "../src/analysis/diagnostics/context.ts";
import { assertSnapshot } from "std/testing/snapshot.ts";

await Parser.init();
const TypeScript = await Parser.Language.load(
  join(testDir, "../grammars/typescript.wasm"),
);

Deno.test("semantic analysis of expose", async (t) => {
  const fileUri = new URL("typegraphs/deno_types.ts", import.meta.url);
  const codeBuf = await Deno.readFile(fileUri);
  const code = new TextDecoder().decode(codeBuf);

  const parser = new Parser();
  parser.setLanguage(TypeScript);

  const tree = parser.parse(code);
  const node = tree.rootNode;
  const ctx = new ModuleDiagnosticsContext(node, fileUri.toString());

  const typegraphDefs = findTypegraphDefinitions(node);
  assertEquals(typegraphDefs.length, 1);

  ctx.checkTypegraph(typegraphDefs[0]);

  await assertSnapshot(t, ctx.diagnostics);
});
