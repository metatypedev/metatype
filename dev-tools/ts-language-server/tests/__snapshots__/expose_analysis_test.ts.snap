export const snapshot = {};

snapshot[`semantic analysis of expose 1`] = `
[
  {
    message: "expected only one parameter for the typegraph builder",
    range: {
      end: {
        character: 27,
        line: 5,
      },
      start: {
        character: 21,
        line: 5,
      },
    },
    severity: 1,
    source: "typegraph",
  },
  {
    message: "unknown type: t.floaty",
    range: {
      end: {
        character: 57,
        line: 18,
      },
      start: {
        character: 47,
        line: 18,
      },
    },
    severity: 1,
    source: "typegraph",
  },
  {
    message: "expected type",
    range: {
      end: {
        character: 60,
        line: 18,
      },
      start: {
        character: 6,
        line: 18,
      },
    },
    severity: 1,
    source: "typegraph",
  },
  {
    message: "Exposed function 'scalarInputType': expected input type to be a struct but got TgType(integer)",
    range: {
      end: {
        character: 46,
        line: 23,
      },
      start: {
        character: 35,
        line: 23,
      },
    },
    severity: 1,
    source: "typegraph",
  },
]
`;
